### PR TITLE
DEX-777:  api for initiating annot identification (including matching set query)

### PR DESCRIPTION
**Pull request by naknomum**
## Pull Request Overview

- Introduces new API to start identification on a single Annotation:  `/api/v1/annotations/identify/GUID` (optionally can pass ES query for _matching-set_ on POST body)
- Sets default algorithm in the event no `idConfigs` value is provided during Sighting submission, so pipeline will complete
- Bonus tweak of json output (related to ident algorithms) of `/api/v1/site-settings/detection`

---
**REST API Updates**

Initiate Identification on a single Annotation.  Can _optionally_ include an Elasticsearch query on the body of the POST.  If not query is provided, the _default_ matching-set query will be used.  (Returns `{ "job_count": N }` with number of ID jobs successfully started.)

```
POST   /api/v1/annotations/identify/GUID
{
    "bool": {
        "filter": [ ...... ],
    }
}
```

### CODE REVIEW

File: app/modules/annotations/models.py [hunk 1]
```diff
@@ -495,3 +495,43 @@ def get_elasticsearch_schema(cls):
         from app.modules.annotations.schemas import AnnotationElasticsearchSchema
 
         return AnnotationElasticsearchSchema
+
+    # this initiates ID pipeline *for this single* annotation.  it is roughly analogous to sighting.ia_pipeline()
```
Comment (hwindsor) [inline app/modules/annotations/models.py:499]: I'm really not sure this should be in the annot module. It seems to be doing the same as the ia_pipeline code in the sightings module on a subset of annots, in this case the subset is one but as this uses all the code in the sightings module to send the request and handle the response, the annot module feels like the wrong place to be initiating it.

File: app/modules/asset_groups/models.py [hunk 2]
```diff
@@ -833,8 +833,16 @@ def remove_annotation(self, annot_guid):
...
                 self.config['encounters'][encounter_num]['annotations'].remove(annot_guid)
 
+    # will default to hotspotter if no idConfigs provided by AGS
     def get_id_configs(self):
-        return self.config.get('idConfigs', [])
+        return self.config.get(
+            'idConfigs',
+            [
+                {
+                    'algorithms': ['hotspotter_nosv'],
```
Comment (hwindsor) [inline app/modules/asset_groups/models.py:842]: This is a nice idea but breaks the rule that there should be no platform specific content in the code and this is definitely zebra specific. 
Instead use ia_config_reader.get(f'_identifiers') to see what is in the IA config and use the first one. 
Also this functionality is also more correct in the sightings model than in the AGS code as that is the place where the algorithms are used.

File: app/modules/annotations/models.py (hunk 1 above)
Comment (hwindsor) [reply app/modules/annotations/models.py:487]: Update: Given the changes I'm currently doing for retries, this functionality now definitely needs to be in the sighting module and not in annotation module.

File: app/modules/asset_groups/models.py (hunk 2 above)
Comment (hwindsor) [reply app/modules/asset_groups/models.py:842]: As I was doing a change in this area too, I put the fix in the sighting code that uses the first identifier in the IA_json:
 https://github.com/WildMeOrg/houston/pull/564/files#diff-e80c56afa47650611e07af0b11959dfa836cad5f07c1bc1d8ffb2c122ee3c2e3R1055-R1062

File: app/modules/ia_config_reader.py [hunk 3]
```diff
@@ -85,7 +85,14 @@ def get_identifiers_list(self, genus_species, ia_class):
     def get_identifiers_dict(self, genus_species, ia_class):
         identifiers = self.get_identifiers_with_links(genus_species, ia_class)
         identifiers_dict = self._resolve_links_to_dict(identifiers)
-        return identifiers_dict
+        # trim the '_identifiers.' prefix off the keys
```
Comment (hwindsor) [inline app/modules/ia_config_reader.py:88]: I'm not clear why you changed this. It doesn't seem related to the other changes in the code. WIP?

Comment (naknomum) [reply app/modules/ia_config_reader.py:88]: just making note what we covered in slack:  this is basically a fix for a bug in the parsing of ia_config data, in order to give frontend the desired json values.  (not truly related to this PR other than it has to do with identification config.)

File: app/modules/asset_groups/models.py (hunk 2 above)
Comment (naknomum) [reply app/modules/asset_groups/models.py]: updated; thanks for the template for this.

File: app/modules/asset_groups/models.py [hunk 4]
```diff
@@ -867,8 +867,19 @@ def remove_annotation(self, annot_guid):
             ):
                 self.config['encounters'][encounter_num]['annotations'].remove(annot_guid)
 
+    # will default to hotspotter if no idConfigs provided by AGS
     def get_id_configs(self):
-        return self.config.get('idConfigs', [])
+        configs = self.config.get('idConfigs')
```
Comment (hwindsor) [inline app/modules/asset_groups/models.py:872]: This should now be reverted entirely and the one in sightings used

File: app/modules/sightings/models.py [hunk 5]
```diff
@@ -697,7 +697,12 @@ def build_identification_request(
         return id_request
 
     def send_identification(
-        self, config_id, algorithm_id, annotation_uuid, annotation_sage_uuid
+        self,
+        config_id,
+        algorithm_id,
+        annotation_uuid,
+        annotation_sage_uuid,
+        matching_set_query_override=None,
```
Comment (hwindsor) [inline app/modules/sightings/models.py:705]: This would be simpler if it was changed to just pass the matching_set_query in all cases rather than the override.

File: app/modules/annotations/models.py [hunk 6]
```diff
@@ -480,3 +483,27 @@ def get_elasticsearch_schema(cls):
         from app.modules.annotations.schemas import AnnotationElasticsearchSchema
 
         return AnnotationElasticsearchSchema
+
+    def send_to_identification(self, matching_set_query=None):
```
Comment (hwindsor) [inline app/modules/annotations/models.py:487]: This is what I said should be in the sightings module. Then it could be used for ia_pipeline, annot ID and encounter ID. 
Is there any reason you have not done so that I'm missing?

File: app/modules/sightings/models.py [hunk 7]
```diff
@@ -1153,30 +1150,48 @@ def ia_pipeline(self):
...
 
         log.debug(
             f'Processing ID for Sighting:{self.guid}, config:{config_id}, algo:{algorithm_id} '
             f'annot:{annotation.guid}, sage_annot:{annotation.content_guid} enc:{annotation.encounter_guid}'
         )
+        assert self.id_configs and config_id >= 0 and config_id < len(self.id_configs)
         if not annotation.content_guid or not annotation.encounter_guid:
             log.warning(f'Skipping {annotation} due to lack of content_guid or encounter')
             return False
         # force this to be up-to-date in index, just to be safe (only do once, hence config_id==0)
```
Comment (hwindsor) [inline app/modules/sightings/models.py:1189]: comment now out of date

File: app/modules/sightings/models.py [hunk 8]
```diff
@@ -719,9 +719,9 @@ def send_identification(
 
         # Message construction has to be in the task as the jobId must be unique
         job_uuid = uuid.uuid4()
-        matching_set_query = matching_set_query_override or self.id_configs[
-            config_id
-        ].get('matching_set')
+        matching_set_query = matching_set_query or self.id_configs[config_id].get(
```
Comment (hwindsor) [inline app/modules/sightings/models.py:722]: not really concerned about this one, I can tweak it later but what I meant was that  matching_set_query should be a mandatory, not optional parameter in this function and in the send_identification task, then it was always set to the correct value in the call to send_identification (which it is)

**Review by hwindsor — APPROVED** [review]
LGTM
